mapie/multi_label_classification.py

```diff
@@ -0,0 +1,795 @@
+                )
+
+                # Split the calculation in two to prevent memory issues
+                batches = [
+                    range(int(n_obs / 2)),
+                    range(n_obs - int(n_obs / 2), n_obs)
+                ]
+                K_R_max = np.zeros((self.n_lambdas, self.n_lambdas))
+                for batch in batches:
+                    nu_batch = nu[batch]
+                    losses_batch = self.risks[batch]
+
+                    nu_batch = np.repeat(
+                        np.expand_dims(nu_batch, axis=2),
+                        self.n_lambdas,
+                        axis=2
+                    )
+                    losses_batch = np.repeat(
+                        np.expand_dims(losses_batch, axis=2),
+                        self.n_lambdas,
+                        axis=2
+                    )
+
+                    R = np.arange(self.n_lambdas) / self.n_lambdas
+                    K_R = np.cumsum(
+                        np.log(
+                            (
+                                1 -
+                                nu_batch *
+                                (losses_batch - R)
+                            ) +
+                            np.finfo(np.float64).eps
+                        ),
+                        axis=0
+                    )
+                    K_R = np.max(K_R, axis=0)
+                    K_R_max += K_R
+
+                r_hat_plus_tronc = self.lambdas[np.argwhere(
+                    np.cumsum(K_R_max > -np.log(delta), axis=1) == 1
+                )[:, 1]]
+                r_hat_plus = np.ones(self.n_lambdas)
+                r_hat_plus[:len(r_hat_plus_tronc)] = r_hat_plus_tronc
+
+        else:
+            r_hat_plus = (n_obs / (n_obs + 1)) * r_hat + (1 / (n_obs + 1))
+
+        return r_hat, r_hat_plus
+
+    def _find_lambda_star(
+        self,
+        r_hat_plus: NDArray,
+        alpha_np: NDArray
+    ) -> NDArray:
+        """Find the higher value of lambda such that for
+        all smaller lambda, the risk is smaller, for each value
+        of alpha.
+
+        Parameters
+        ----------
+        r_hat_plus : NDArray of shape (n_lambdas, )
+            Upper bounds computed in the get_r_hat_plus method.
+        alphas : NDArray of shape (n_alphas, )
+            Risk levels.
+
+        Returns
+        -------
+        NDArray of shape (n_alphas, )
+            Optimal lambdas which control the risks for each value
+            of alpha.
+        """
+
+        if len(alpha_np) > 1:
+            alphas_np = alpha_np[:, np.newaxis]
+        else:
+            alphas_np = alpha_np
+
+        bound_rep = np.repeat(
+            np.expand_dims(r_hat_plus, axis=0),
+            len(alphas_np),
+            axis=0
+        )
+        bound_rep[:, np.argmax(bound_rep, axis=1)] = np.maximum(
+            alphas_np,
+            bound_rep[:, np.argmax(bound_rep, axis=1)]
+        )
+        lambdas_star = np.argmin(
```

Comment:
Use the stored values self.lambdas : much more compact and easier to understand

Reply:
done